### PR TITLE
Use JNA to Speed up Snapshot Cache File Creation

### PR DESCRIPTION
**Pull request by original-brownbear**
Use JNA to speed up snapshot cache file creation. Do this in `:server`
to bypass the security filter and move necessary bits of code to `:server`
to enable the logic.
Fall back to trying to create the file by writing zeros if anything except for
the step determining free disk space fails.

Manually verified via `du` that this works as expected and actually allocates hundreds of GB in a few `ms`.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java [hunk 1]
```diff
@@ -260,4 +269,39 @@ static void tryInstallSystemCallFilter(Path tmpFile) {
             logger.warn("unable to install syscall filter: ", e);
         }
     }
+
+    @SuppressForbidden(reason = "need access to fd on FileOutputStream")
+    static void fallocateSnapshotCacheFile(Environment environment, long fileSize) throws IOException {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java:274]: It's a little unfortunate to leak that much logic into this class, but I figured abstracting things away further to isolate the `fallocate` as much as possible wasn't really worth the extra complication (IMO).

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 2]
```diff
@@ -132,6 +133,26 @@
 
     public static final String UPDATE_SNAPSHOT_STATUS_ACTION_NAME = "internal:cluster/snapshot/update_snapshot_status";
 
+    public static final String SEARCHABLE_SNAPSHOT_SETTINGS_PREFIX = "xpack.searchable.snapshot.shared_cache.";
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:136]: this is the prefix for the shared cache, not for all SEARCHABLE_SNAPSHOT_SETTINGS

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 3]
```diff
@@ -132,6 +133,26 @@
...
+
+    public static final Setting<ByteSizeValue> FROZEN_CACHE_RANGE_SIZE_SETTING = Setting.byteSizeSetting(
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:138]: let's name this SHARED_CACHE_RANGE_SIZE_SETTING

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 4]
```diff
@@ -132,6 +133,26 @@
...
+        FROZEN_CACHE_RANGE_SIZE_SETTING,
+        Setting.Property.NodeScope
+    );
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_SIZE_SETTING = Setting.byteSizeSetting(
+        SEARCHABLE_SNAPSHOT_SETTINGS_PREFIX + "size",
+        ByteSizeValue.ZERO,
+        Setting.Property.NodeScope
+    );
+
+    public static final String CACHE_FILE_NAME = "snap_cache";
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: ++ Renamed to `shared_snapshot_cache` :)

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java]: ++ renamed.

**Comment by original-brownbear** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java [hunk 5]
```diff
@@ -260,4 +269,39 @@ static void tryInstallSystemCallFilter(Path tmpFile) {
...
+        if (Constants.LINUX == false) {
+            logger.debug("not trying to create a shared cache file using fallocate on non-Linux platform");
+            return;
+        }
+        Path cacheFile = SnapshotUtils.findCacheSnapshotCacheFilePath(environment, fileSize);
+        if (cacheFile == null) {
+            throw new IOError(new IOException("Could not find a directory with adequate free space for cache file"));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java:281]: why wrap in IOError?

**Review by ywelsch — APPROVED** [review]
LGTM. Would like to see a review from one more person before merge

File: server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java (hunk 5 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/bootstrap/JNANatives.java]: Remove that wrapping, sorry for the noise, this was a left-over from some debugging.

File: server/src/main/java/org/elasticsearch/bootstrap/Natives.java [hunk 6]
```diff
@@ -132,4 +134,20 @@ static boolean isSystemCallFilterInstalled() {
...
+    /**
+     * On Linux, this method tries to create the searchable snapshot frozen cache file using fallocate if JNA is available. This enables
+     * a much faster creation of the file than the fallback mechanism in the searchable snapshots plugin that will pre-allocate the cache
+     * file by writing zeros to the file.
+     *
+     * @throws IOException on failure to determine free disk space for a data path
+     */
+    public static void tryCreateCacheFile(Environment environment, long fileSize) throws IOException {
+        if (JNA_AVAILABLE == false) {
+            logger.warn("cannot use fallocate to create cache file because JNA is not available");
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/bootstrap/Natives.java:147]: nit: I know we are not consistent in starting log messages with uppercase or not but we're using both ways in this PR

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 7]
```diff
@@ -132,6 +133,26 @@
...
+    public static final String SHARED_CACHE_SETTINGS_PREFIX = "xpack.searchable.snapshot.shared_cache.";
+
+    public static final Setting<ByteSizeValue> SHARED_CACHE_RANGE_SIZE_SETTING = Setting.byteSizeSetting(
+        SHARED_CACHE_SETTINGS_PREFIX + "range_size",
+        ByteSizeValue.ofMb(16),                                 // default
+        Setting.Property.NodeScope
+    );
+    public static final Setting<ByteSizeValue> SNAPSHOT_CACHE_REGION_SIZE_SETTING = Setting.byteSizeSetting(
+        SHARED_CACHE_SETTINGS_PREFIX + "region_size",
+            SHARED_CACHE_RANGE_SIZE_SETTING,
```
Comment (tlrx) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:145]: nit ident

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick + Tanguy!

**Comment by joegallo** [comment]
Reverted in 4c2e9a6d99849ab9970e18f747b7229a73ded5cc because it broke local development on a mac. (Sorry!)

**Pull request merged**